service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/disk/DetachAzureDiskFromVmStep.java

```diff
@@ -0,0 +1,190 @@
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.app.configuration.external.AzureConfiguration;
+import bio.terra.workspace.common.exception.AzureManagementExceptionUtils;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.resource.controlled.cloud.azure.DeleteAzureControlledResourceStep;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import com.azure.core.management.exception.ManagementException;
+import com.azure.resourcemanager.compute.ComputeManager;
+import com.azure.resourcemanager.compute.models.Disk;
+import com.azure.resourcemanager.compute.models.VirtualMachine;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This steps works as a companion step for a disk deletion process. It provides an additional
+ * functionality of detaching data disk from a virtual machine before disk deletion. It is not
+ * possible to delete a disk without detaching it.
+ */
+public class DetachAzureDiskFromVmStep extends DeleteAzureControlledResourceStep {
+  private static final Logger logger = LoggerFactory.getLogger(DetachAzureDiskFromVmStep.class);
+
+  private final CrlService crlService;
+  private final AzureConfiguration azureConfig;
+  private final ControlledAzureDiskResource resource;
+
+  public DetachAzureDiskFromVmStep(
+      AzureConfiguration azureConfig, CrlService crlService, ControlledAzureDiskResource resource) {
+    this.crlService = crlService;
+    this.azureConfig = azureConfig;
+    this.resource = resource;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final AzureCloudContext azureCloudContext =
+        DeleteAzureDiskFlightUtils.getAzureCloudContext(context);
+    final String diskResourceId =
+        DeleteAzureDiskFlightUtils.getAzureDiskResourceId(azureCloudContext, resource);
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+    try {
+      var disk = computeManager.disks().getById(diskResourceId);
+      if (!disk.isAttachedToVirtualMachine()) {
+        logger.info(
+            "Disk {} in workspace {} is not attached to vm.",
+            resource.getResourceId(),
+            resource.getWorkspaceId());
+        return StepResult.getStepResultSuccess();
+      }
+      var vm = getVirtualMachine(computeManager, disk.virtualMachineId());
+      if (vm.isEmpty()) {
+        // this is unlikely event - disk is attached, but the vm cannot be found; let's retry
+        logger.warn(
+            "Disk {} is attached to vm in workspace {}, but vm cannot be found. Retrying.",
+            resource.getResourceId(),
+            resource.getWorkspaceId());
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY);
+      }
+      var attachedDisk =
+          vm.get().dataDisks().entrySet().stream()
```

Comment:
In this context `entrySet` is a part of HashMap implementation and it could not be null. According to the Java SDK `dataDisk()` also returns empty collection and it is unlikely that SDK behavior in this context will change. But I agree that we need to protect ourselves.